### DISCUSSION TITLE
Loading of CSS & JS files.

### DISCUSSION THREAD

**Discussion by bislewl** [zencart/zencart#6939] [Ideas]
CSS/JS loader seems to help with the minify, and not loading unneeded files. However it is very difficult for most beginners to use, and isn't friendly with most templates. Anyone have thoughts about modifying the core for CSS/JS loading. I'm thinking perhaps instead of auto_loader files. Making it admin configurable. and simpler for people to use. (select which files load when). Adding the ability to minify the CSS/JS files. Also I have seen many sites using 3 OR MORE versions of jQuery and having JS issues, so not sure how we can solve that.

**Comment by IronLady** [comment]
Maybe using hooks like wordpress did. Template or module can enqueue its own css/js.
I'd love to see this feature in zen cart.

**Comment by drbyte** [comment]
v160 already has a number of things built-in to accommodate these. 

I'd be interested in seeing what difficulties you encounter when trying to adapt CSS/JS Loader to work with new core code.  I too think the CSS/JS Loader addon is bloated and hard to use for most people, and lends itself to over-duplication.

**Comment by DivaVocals** [comment]
> CSS/JS loader seems to help with the minify, and not loading unneeded files. However it is very difficult for most beginners to use, and isn't friendly with most templates. Anyone have thoughts about modifying the core for CSS/JS loading. I'm thinking perhaps instead of auto_loader files. Making it admin configurable. and simpler for people to use. (select which files load when). Adding the ability to minify the CSS/JS files. Also I have seen many sites using 3 OR MORE versions of jQuery and having JS issues, so not sure how we can solve that.

# 

> I too think the CSS/JS Loader addon is bloated and hard to use for most people, and lends itself to over-duplication.

I have made no secret my disdain for the CSS/JS loader. I am not opposed to the BENEFIT it is supposed to bring.. What I've always objected to is the fact that it's not just difficult for beginners, it's difficult for folks who are NOT beginners too.. I find that the CSS/JS Loader is poorly documented, not user friendly, lacking in ANY admin level controls so that you can strategically manage how the minification is applied. In it's current form, it's either ON or OFF.. there's no in between.. WTH??? and here's ALL the instruction you get in the readme for when things go wrong (which for many is usually as soon as you turn this very non-configurable module ON..):

> If you had embedded JavaScript, CSS, or meta data into your backed-up html_header.php file, either move these lines of code to includes/templates/YOUR_TEMPLATE/jscript/jscript_sitewide.php (name of file can be anything and will load site-wide by Zen Cart), or build an auto loader for CSS JS Loader in includes/templates/YOUR_TEMPLATES/auto_loaders/ (recommended if using shared libraries like jQuery). 

Now I have been working with Zen Cart since 2006, and **I** have to refresh my memory on creating an auto-loader.. What exactly is a NOVICE DIY shop owner supposed to do with this bit of NON information??? :laughing: 

Finally I NEVER like the fact that nearly every single one of a popular contributor's modules FORCES you to install the CSS/JS Loader whether you want it or not.. With all due respect, I personally find that the "we know what's good for you better than you do" stance taken by this contributor is a bit akin to the "Borg" in Star Trek (you know.. that whole "resistance is futile" POV :laughing: ). 

I yanked the CSS/JS Loader out of the TPP module because bundling one module inside another a REALLY bad idea unless you intend to maintain BOTH (the person who added this into TPP did not). For those of us left behind to deal with the fallout of this galacticly BAD mashup idea and help novice shop owners getting their sites working again, I found that the inclusion of this intrusive module created more HARM than good.. I'm not suggesting that I ever wanted to deny anyone the right to use TPP and the CSS/JS Loader together, but I see them as separate installs since TPP functions JUST fine without the CSS/JS Loader.. If those who support the CSS/JS/ Loader want to help that shopowner get TPP working with the CJJ/JS Loader then they should handle that..

However, I digress.. :laughing: I don't want to come across as if I am slamming the CSS/JS Loader.. I'm not slamming the module's INTENT, just the method in which it has been supported, implemented and distributed.. I take issue with these things.. 'tis all.. that ALL said, I would **LOVE** to see a minification module that is easy to manage/configure, and doesn't require novice shop owners to know how to create an "auto loader" file JUST so the minification doesn't break their site.. 

I'd like to see the following features in any minification built into Zen Cart:
- CONFIGURABLE.. Don't FORCE this on a shopowner.. PLEASE don't take the "we know what's good for you better than you do" POV. If they CHOOSE to not turn on minification, then they should have that CHOICE..
  - Enable JavaScript Minification: Check the option if you want to minify Javascriipt.
  - Enable CSS Minification: Check the option if you want to minify CSS.
  - Enable HTML Minification: Check the checkbox if you want to minify HTML.
  - JavaScript files to exclude from minify (line delimited): Enter the name of JavaScript files that you don’t want to minify. Name should appear one per line.
  - CSS files to exclude from minify (line delimited): Functions as above.
- Provide an option to choose to load certain JS and/or CSS libraries from the CDN of their choice with a fallback to a local version if needed (I've seen similar features in WordPress plugins in this class)

> Maybe using hooks like wordpress did. Template or module can enqueue its own css/js.
> I'd love to see this feature in zen cart. 

Yes, but there needs to be some guidelines/standards.. even in WordPress there are still a few STRAY plugin authors who enqueue css/js that may be in conflict with the WordPress core or even other plugins.. I'm thinking specifically of some of the Bootstrap or Foundation plugins which don't give you options to use Bootstrap or Foundation libraries from other plugins so you don't do something silly like load Bootstrap or Foundation libraries twice. </ endpost > :laughing: 

_See what happens when I finally get some time on my hands_ :smiley_cat:

**Comment by drbyte** [comment]
To elaborate on my previous note:

> v160 already has a number of things built-in to accommodate these.

You can see the code for this in the /includes/modules/tpl_css_js_generator.php script:

## Interception
- a notifier at the start to completely change the defined css-js handler if one wishes to completely bypass the built-in one

## BODY Classes
- BODY classes dynamically added based on page-name, current-product-id, current-mfg-id, current-category-id, current-cPath, current-language
  ... and a notifier to allow complete alteration/override of all those

## Framework injected css, js
- notifier to set which CSS framework is in use (in case elsewhere something needs to be set based on the named framework)

## Stylesheet files

Stylesheet filenames are loaded in this order, if present:
1. all stylesheets for your selected CSS framework, or whatever were set via the notifier hook (if any), followed by:
2. stylesheet.css
3. stylesheet-responsive.css
4. font.css
5. inline_en_stylesheet.css
6. inline_pagename.css
7. inline_en_pagename.css
8. inline_c_cpath.css
9. inline_en_c_cPath.css
10. inline_m_mfgId.css
11. inline_en_m_mfgId.css
12. inline_p_productId.css
13. inline_en_p_productId.css
14. print*.css

De-duplication is done to ensure duplicates/conflicts are managed when the framework dictates certain files, and the template has same files.

CSS `<link>` tags are assigned an `id` so that javascripts can enable/disable them if necessary.

Naturally we encourage people to not use inline styling, wherever possible, but the feature is present to support the flexibility oft-requested by the community at-large.

## JS files
- jscript_top_xxxx.php and jscript_top_xxxxx.js files are loaded at the top of the page (in HEAD)
- other jscript_xxxxx.php and jscript_xxxxx.js files are loaded at the bottom of the page (below footer) to prevent blocking
- De-duplication is done to ensure duplicates/conflicts are managed when the framework dictates certain files, and the template has same files.

**Comment by drbyte** [comment]
And, of course, with notifiers, tons of minification can be done via external tools, if desired, with no need to mangle any core code.

**Comment by DivaVocals** [comment]
@drbyte this sounds pretty awesome.. I think.. :laughing: .. speaking from the POV of one of the "non-coders" in the bunch I see the term "notifiers" used, but I am fully unclear what they mean or how one can use these little golden nuggets.. Can you help my non-coding brain understand what a notifier is?? :smile;

**Comment by drbyte** [comment]
One of the things I forgot to mention in my earlier post to this discussion is that the choice of "CSS Framework" is also already built-in, and is template-specific, chosen by the template-designer: Simply set  an appropriate value for `$template_framework` in the `template_info.php` file, and then use the observers mentioned above to intercept the default CSS/JS operations and thus determine when/where to output the corresponding CSS/JS code in the HTML sent to the browser.  

There are various ways to do it, but this is the concept originally intended when we redesigned that part of the template code.

Notifiers/Observers explained: http://www.zen-cart.com/wiki/index.php/Developers_API_Tutorials#Observer_Class

**Comment by mesnitu** [comment]
Glad to see that the body will get all those classes previously only available with strange css files.

**Comment by lat9** [comment]
Re: The javascript files' loading.  Since all `jscript_*` files are now loading in the `<head>`, I'll suggest for backwards-compatibility that we use the naming convention of `jscript_footer_*.php|js` to identify files that should be loaded (alphabetically) in the footer.

I'm also unclear as to the ordering of CSS files when there is a mixture of those that start with `style` and those that start with `inline`.  Is there an interspersing of `<link>` tags to load a file and `<style>` tags to include those inlined ones?

**Reply by torvista** [reply]
As far as I can see,  
files various**.css** get inserted as links to the file   
files style***.php**/site_specific_styles**.php** get included inline in their own style tags.

**Comment by torvista** [comment]
I've been wading through a CSS-JS loader implementation over the weekend, liberally scattering debug info all over the place to get to the bottom of it and its problems and am constantly in two minds..."this could be good/should be in core" / "this is too complicated for users and needs to go in the bin."

**Reply by torvista** [reply]
FYI
CSS-JS Loader by default loads all the javascript in the footer and some css in the head, unless marked as defer, in which case it goes in 
 the footer.
I modified this for javascript so jscript_head_xxx goes in the head and all the other javascript goes in the footer by default.
Other functionalities from CSS-JS Loaded are
1. choose the page on which things are loaded.
2. choose the order in which things are loaded.
3. Minify the scripts and concatenate them into one link

Not making the case for this add-on, just noting the functions to bear in mind.

**Reply by scottcwilson** [reply]
"this is too complicated for users and needs to go in the bin." 

@drbyte has stated this before.
